### PR TITLE
Fixed clicking on the min value when there is no value yet

### LINKED ISSUE #6504: Impossible to tap the first point on the axis in the range widget
#### ODK Collect version
the store version 2024.2.4, the 2024.3.0 Beta 8
#### Android version
10, 14
#### Device used
Redmi 9T, Pixel 7a
#### Problem description
It isn’t possible to tap the first point on the axis in the range widget unless other point was selected before.

https://github.com/user-attachments/assets/506319c8-55f4-4f86-9b75-42cbed34308c

#### Steps to reproduce the problem
1. Open range widget form.
[rangewidget.xml.txt](https://github.com/user-attachments/files/17717440/rangewidget.xml.txt)
2. Try to tap the first point on the axis
#### Expected behavior
It should be possible to select the first point on the axis

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #6504 

#### Why is this the best possible solution? Were any other approaches considered?
The issue arises because all our question types support a "no answer" state, whereas sliders lack this concept. To address this, we adjusted the slider's appearance but couldn't set its value outside the specified range (or provide no answer at all). For example, if the range was 1–10, the value defaulted to 1 (the minimum) when no answer was provided. As a result, clicking the minimum value didn't trigger a click event since consecutive clicks on the same value don't generate multiple value-changed events.

To resolve this, I implemented a separate interface to listen for touch events and trigger the missing event when no answer has been provided yet.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
As mentioned above, the issue occurred when clicking on the minimum value with no answer provided. Please test this scenario as well as clicking on other values to ensure everything works correctly.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] added any new strings with date formatting to `DateFormatsTest`
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

File: androidtest/src/main/java/org/odk/collect/androidtest/SliderExt.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package org.odk.collect.androidtest
+
+import android.view.MotionEvent
+import com.google.android.material.slider.Slider
+
+fun Slider.clickOnStep(step: Float) {
```
Comment (seadowg) [inline androidtest/src/main/java/org/odk/collect/androidtest/SliderExt.kt]: I think this should probably live in `testshared` instead as it's not the kind of thing that I feel should be `androidx.test:core` (which is what `androidtest` is really there for).

File: collect_app/src/test/java/org/odk/collect/android/widgets/range/RangeDecimalWidgetTest.java [hunk 2]
```diff
@@ -202,6 +203,17 @@ public void everyTriggerWidgetShouldHaveCheckboxWithUniqueID() {
         assertThat(widget1.slider.getId(), not(equalTo(widget2.slider.getId())));
     }
 
+    @Test
+    public void changingSliderValueToTheMinOneWhenSliderHasNoValue_setsTheValue() {
+        RangeDecimalWidget widget = createWidget(promptWithQuestionDefAndAnswer(rangeQuestion, null));
+        widget.slider.measure(100, 10);
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/widgets/range/RangeDecimalWidgetTest.java]: We could add a helper for the `measure`/`layout` boilerplate to `SliderExt` as well.

**Review by seadowg — CHANGES_REQUESTED** [review]
I don't think I understand the implementation. I can remove the `enable`/`disable` calls and no tests fail, so maybe I'm missing some edge cases that need tests?

File: androidtest/src/main/java/org/odk/collect/androidtest/SliderExt.kt (hunk 1 above)
Comment (grzesiek2010) [reply androidtest/src/main/java/org/odk/collect/androidtest/SliderExt.kt]: Done.

File: collect_app/src/test/java/org/odk/collect/android/widgets/range/RangeDecimalWidgetTest.java (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/test/java/org/odk/collect/android/widgets/range/RangeDecimalWidgetTest.java]: It turns out `widget.slider.measure(100, 10);` is not needed so I removed it.

**Comment by grzesiek2010** [comment]
> I don't think I understand the implementation.

Let me explain:

1. When the range question type has no value yet, we programmatically make the slider appear as though it has no value. However, since the slider doesn't inherently support this concept, its value is set to the minimum value under the hood (set but not displayed).  
2. If you try to click on the minimum value, the `onValueChanged` event is not triggered because it assumes no change has occurred.  
3. This was the issue: if you first clicked on any value other than the minimum, subsequent clicks on the minimum value would trigger the event since it now recognized a change in value.  
4. To resolve this, I added a custom listener to detect clicks on the slider when it is in a "disabled" state (i.e., no value set yet).  
5. This listener operates via an `OnTouchListener` in a way that ensures it is triggered only when necessary. Specifically:  
   - If you click on a value other than the minimum, `onValueChanged` is already called, and the slider is marked as enabled, so the additional listener is not invoked.  
   - If you click on the minimum value, `onValueChanged` is not called, leaving the slider in a "disabled" state. In this case, the custom listener detects the interaction and handles it appropriately.

> I can remove the enable/disable calls and no tests fail, so maybe I'm missing some edge cases that need tests?

This is because I initially added tests only to cover the bug. However, the state is also used to prevent updating the value twice. As mentioned above, if the slider is already enabled, there’s no need to call the listener. I added new tests to cover that.

**Review by seadowg — CHANGES_REQUESTED** [review]
I think there are two things I'm finding confusing here:

1. We've ended up with two different enabled/disabled concepts for `Slider`
2. There are two listeners that the widgets have to interact with two get value changes

I'm wondering if we could just introduce a single custom listener to solve this and just not use `Slider.OnChangeListener` at all. This listener could have the method signature as `onValueChange` (except for `fromUser` which we wouldn't need), but would just always propagate the value being selected (even if there was no real change). We'd end up with something like this in `TrackingTouchSlider`:

```kotlin
setOnTouchListener { v: View, event: MotionEvent ->
    val action = event.action
    when (action) {
        MotionEvent.ACTION_DOWN -> v.parent.requestDisallowInterceptTouchEvent(
            true
        )

        MotionEvent.ACTION_UP -> {
            v.parent.requestDisallowInterceptTouchEvent(false)
            listener.onValueChanged(this, value)
        }
    }
    v.onTouchEvent(event)
    true
}
```

Then, when the widget is cleared, you could just call "reset" it to the minimum value.

**Comment by grzesiek2010** [comment]
@seadowg you were right this could have been simplified.

File: collect_app/src/main/java/org/odk/collect/android/views/TrackingTouchSlider.kt [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    override fun onStopTrackingTouch(slider: Slider) {
+        isTrackingTouch = false
+    }
+
+    @SuppressLint("RestrictedApi")
+    override fun onStartTrackingTouch(slider: Slider) {
+        isTrackingTouch = true
+    }
+
+    fun enable() {
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/views/TrackingTouchSlider.kt]: > We've ended up with two different enabled/disabled concepts for Slider

I don't think [this feedback](https://github.com/getodk/collect/pull/6524#pullrequestreview-2458187092) got addressed. To expand on this, it's confusing to have `enable`/`disable` because `View` already has a `isEnabled` property. My thinking that we could have something like a single `reset` method instead that does basically what `disable` does (and is called by a widget when the answer is cleared) and that the `enable` part could be handled entirely within `TrackingTouchSlider` when a user change happens.

**Review by seadowg — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/views/TrackingTouchSlider.kt (hunk 3 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/views/TrackingTouchSlider.kt]: Done.

**Review by seadowg — APPROVED** [review]

**Comment by seadowg** [comment]
@grzesiek2010 didn't want to merge this as you hadn't marked it ready for another review yet, but I'm assuming that was a mistake. Go ahead and merge whenever if so!

**Comment by dbemke** [comment]
**Tested with Success!**

Verified on devices with Androids: 10, 14

**Verified cases:**
- issue #6504 doesn't occur
- forms with range widget
- removing the selected point on the axis, going to the hierarchy and back
- RTL
- don't keep activities enabled/disabled
